torch/_library/abstract_impl.py

```diff
@@ -0,0 +1,197 @@
+
+    def create_dynamic_size(self, *, min=2, max=None) -> torch.SymInt:
+        """Constructs a new symint (symbolic int) representing a data-dependent value.
+
+        This is useful for writing the abstract implementation (which is necessary
+        for torch.compile) for a CustomOp where an output Tensor has a size
+        that depends on the data of the input Tensors.
+
+        Args:
+            min (int): A statically known inclusive lower bound for this symint.
+                min must be at least 2 due to implementation details of
+                torch.compile. Default: 2.
+            max (Optional[int]): A statically known inclusive upper bound for this
+                symint. Default: None
+
+        .. warning:
+
+            It is important that the ``min`` and ``max`` (if not None) values are set
+            correctly, otherwise, there will be undefined behavior under
+            torch.compile. The default value of ``min`` is 2 due to torch.compile
+            specializing on 0/1 sizes.
+
+            You must also verify that your implementation on concrete Tensors
+            (e.g. CPU/CUDA) only returns Tensors where the size that corresponds
+            to the symint also has respects these constraint.
+            The easiest way to do this is to add an assertion in the CPU/CUDA/etc
+            implementation that the size follows these bounds.
+
+        Example::
+
+            >>> # Let's assume we've registered an operator,
+            >>> # mylibrary::custom_nonzero, with data-dependent output shape
+            >>> @custom_nonzero.impl_abstract():
+            >>> def custom_nonzero_abstract(x):
+            >>>     # Number of nonzero-elements is data-dependent
+            >>>     ctx = torch.library.get_ctx()
+            >>>     nnz = ctx.create_unbacked_symint()
+            >>>     shape = [x.dim(), nnz]
+            >>>     result = x.new_empty(shape, dtype=torch.long)
+            >>>     return result
+
+        """
+        if (
+            self._shape_env is None
+            or not self._shape_env.allow_dynamic_output_shape_ops
+        ):
+            raise torch._subclasses.fake_tensor.DynamicOutputShapeException(self._op)
+
+        if isinstance(min, torch.SymInt) or isinstance(max, torch.SymInt):
+            raise ValueError(
+                f"ctx.create_dynamic_size(min={min}, max={max}): expected "
+                f"min and max to be statically known ints but got SymInt. "
+                f"This is not supported."
+            )
+
+        if min < 2:
+            raise ValueError(
+                f"ctx.create_dynamic_size(min={min}, ...): expected min to be "
+                f"greater than or equal to 2. PyTorch only supports new "
+                f"data-dependent sizes of >= 2"
+            )
+
+        result = self._shape_env.create_unbacked_symint()
+        torch.fx.experimental.symbolic_shapes.constrain_range(result, min=2, max=max)
```

Comment:
@ezyang I never really understood, but, what happens if someone tries to torch.compile(dynamic=True) a call to custom::nonzero that actually returns a Tensor of size (1, 2) (where the symint represents the 1)?

Does this result in a runtime assertion, and if so, where? (inductor?)

Reply:
So, we're fixing this right now in https://github.com/pytorch/pytorch/pull/108372 but yes, as of today, if runtime assertions were working (they only work in export right now) you'd end up with a `>= 2` assertion in the graph which is wrong.

The fixed version of this is that we compile time assume >= 2, but the runtime assert we generate is >= 0 and all is well. This is for the constrain_as_size combinator, which we should be encouraging people to use. Actually, with new improvements to implicit size assumptions, this isn't even necessary if you send the unbacked symint straight to a constructor.

Reply:
Yeah, you should use `constrain_as_size` here

Reply:
That is a WIP PR, so I think I am going to land https://github.com/pytorch/pytorch/pull/109912 first, and then we change it to use `sym_constrain_range_for_size` in that PR